### PR TITLE
Homework 5 task_1 oop_1 done

### PR DESCRIPTION
**Pull request by ivanzuev78**
Homework 5 task_1 oop_1 done

### CODE REVIEW

File: HomeWork_5/oop_1.py [hunk 1]
```diff
@@ -4,57 +4,76 @@
...
+    """
+         Homework принимает на вход 2 атрибута: текст задания и количество дней
+    на это задание
+    Атрибуты:
+        text - текст задания
+        deadline - хранит объект datetime.timedelta с количеством
+        дней на выполнение
+        created - c точной датой и временем создания
+    Методы:
+        is_active - проверяет не истело ли время на выполнение задания,
```
Comment (Cjkjvfnby) [inline HomeWork_5/oop_1.py:26]: `истело` -> `истлело`

File: HomeWork_5/oop_1.py [hunk 2]
```diff
@@ -4,57 +4,76 @@
...
+        возвращает boolean
+    """
+
+    def __init__(self, task_text: str, days_to_do: int):
+        self.text = task_text
+        self.deadline = datetime.timedelta(days_to_do)
```
Comment (Cjkjvfnby) [inline HomeWork_5/oop_1.py:32]: It is better to use keyword argument here.

File: HomeWork_5/oop_1.py [hunk 3]
```diff
@@ -4,57 +4,76 @@
...
+        do_homework - принимает объект Homework и возвращает его же,
+        если задание уже просрочено, то печатет 'You are late' и возвращает None
+    """
+
+    def __init__(self, first_name: str, last_name: str):
+        self.last_name = last_name
+        self.first_name = first_name
 
+    @staticmethod
+    def do_homework(homework: Homework) -> Union[Homework, None]:
```
Comment (Cjkjvfnby) [inline HomeWork_5/oop_1.py:55]: Check Optional in typing module

File: HomeWork_5/oop_1.py [hunk 4]
```diff
@@ -4,57 +4,76 @@
...
+        if homework.is_active():
+            return homework
+        print("You are late")
```
Comment (Cjkjvfnby) [inline HomeWork_5/oop_1.py:58]: If a function returns something it is good practice to explicitly return None

File: HomeWork_5/test_oop_1.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import pytest
+
+from HomeWork_5.oop_1 import *
+
+
+@pytest.fixture
+def teacher():
+    teacher = Teacher("Daniil", "Shadrin")
+    yield teacher
```
Comment (Cjkjvfnby) [inline HomeWork_5/test_oop_1.py:9]: I'd recommend inlining the variable.  Here and in other cases.

File: HomeWork_5/test_oop_1.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    assert expired_homework.is_active() is True
+
+
+def test_homework_deadline():
+    homework = Teacher.create_homework("task", 5)
+    assert homework.deadline == datetime.timedelta(5)
+
+
+def test_student_do_homework_in_time(student):
+    expired_homework = Teacher.create_homework("Learn functions", 1)
```
Comment (Cjkjvfnby) [inline HomeWork_5/test_oop_1.py:40]: Are you sure it is expired?

File: HomeWork_5/test_oop_1.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+def outdated_homework(monkeypatch):
+    outdated_homework = Teacher.create_homework("Learn functions", 1)
+    new_time = datetime.datetime.now() + datetime.timedelta(2)
+
+    class new_datetime:
+        @classmethod
+        def now(cls):
+            return new_time
+
+    monkeypatch.setattr(datetime, "datetime", new_datetime)
```
Comment (Cjkjvfnby) [inline HomeWork_5/test_oop_1.py:54]: You can mock function directly, does not need to do it with the whole module

```python
monkeypatch.setattr(datetime.datetime, "now", lambda: new_time)
```

File: HomeWork_5/test_oop_1.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def test_homework_deadline():
+    homework = Teacher.create_homework("task", 5)
+    assert homework.deadline == datetime.timedelta(5)
+
+
+def test_student_do_homework_in_time(student):
+    expired_homework = Teacher.create_homework("Learn functions", 1)
+    assert student.do_homework(expired_homework) is expired_homework
```
Comment (Cjkjvfnby) [inline HomeWork_5/test_oop_1.py:41]: What else can we check for `do_homework`?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Looks good, just minor comments.

File: HomeWork_5/test_oop_1.py (hunk 5 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py:9]: Done

File: HomeWork_5/oop_1.py (hunk 1 above)
Comment (ivanzuev78) [reply HomeWork_5/oop_1.py:26]: I suppose `истекло`

File: HomeWork_5/oop_1.py (hunk 2 above)
Comment (ivanzuev78) [reply HomeWork_5/oop_1.py:32]: Done

File: HomeWork_5/oop_1.py (hunk 4 above)
Comment (ivanzuev78) [reply HomeWork_5/oop_1.py:58]: I will know. Done.

File: HomeWork_5/test_oop_1.py (hunk 6 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py:40]: I mixed this word up with the "expected".

File: HomeWork_5/test_oop_1.py (hunk 7 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py:54]: I tried to do something like this, but it doesn't works.
```
>       monkeypatch.setattr(datetime.datetime, "now", lambda: new_time)
E       TypeError: can't set attributes of built-in/extension type 'datetime.datetime'
```

File: HomeWork_5/test_oop_1.py (hunk 8 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py:41]: I suppose the deadline.

File: HomeWork_5/test_oop_1.py (hunk 7 above)
Comment (Cjkjvfnby) [reply HomeWork_5/test_oop_1.py:53]:
```
a = []  # datatime
b = a  # reference in you module to datetime (during import)
c = []   
a = c   # monkeypath datetime, but you module already has a link to original a.
```
Try something like this. You need to patch it in your module, not in the original module.
```
import HomeWork_5.oop_1 

monkeypatch.setattr(oop_1.datetime.datetime, "now", lambda: new_time) 
```

File: HomeWork_5/test_oop_1.py (hunk 8 above)
Comment (Cjkjvfnby) [reply HomeWork_5/test_oop_1.py]: Does this function have side effects?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Please check my last comments.

File: HomeWork_5/test_oop_1.py (hunk 7 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py:53]: I suppose that is the answer:
https://stackoverflow.com/questions/4481954/trying-to-mock-datetime-date-today-but-not-working

File: HomeWork_5/test_oop_1.py (hunk 8 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py]: To tell the truth, I can't find side effects.

Comment (Cjkjvfnby) [reply HomeWork_5/test_oop_1.py]: What is the definition of side effects?  What is common examples?

File: HomeWork_5/test_oop_1.py (hunk 7 above)
Comment (Cjkjvfnby) [reply HomeWork_5/test_oop_1.py:52]: Do you trust me or some strangers on the Internet?

Did you read all comments?
https://stackoverflow.com/a/20034516

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Please resolve the comments

File: HomeWork_5/test_oop_1.py (hunk 8 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py]: I didn' know about this meaning: https://en.wikipedia.org/wiki/Side_effect_(computer_science)
Example side effects include modifying a non-local variable, modifying a static local variable.
I suppose `student.do_homework(expired_homework)` can change `expired_homework`
I'll learn about side effects more.

File: HomeWork_5/test_oop_1.py (hunk 7 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py:52]: It breakdown my brain.

File: HomeWork_5/test_oop_1.py (hunk 8 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py]: May be we should check, that after `do_homework(homework)` homework is the same?

Comment (Cjkjvfnby) [reply HomeWork_5/test_oop_1.py]: > Example side effects include modifying a non-local variable, modifying a static local variable, modifying a mutable argument passed by reference, performing I/O or calling other side-effect functions.

Which of this cases is present in the function?

File: HomeWork_5/test_oop_1.py [hunk 9]
```diff
@@ -38,22 +40,25 @@ def test_student_do_homework_in_time(student):
     homework = Teacher.create_homework("Learn functions", 1)
     assert student.do_homework(homework) is homework
 
+    hw_copy = copy.copy(homework)
+    assert hw_copy.text == homework.text
+    assert hw_copy.deadline == homework.deadline
+    assert hw_copy.created == homework.created
```
Comment (Cjkjvfnby) [inline HomeWork_5/test_oop_1.py:46]: This looks like a test of the `copy.copy`.   It is already done here https://github.com/python/cpython/blob/master/Lib/test/test_copy.py#L160.  Why do we need an extra test for this class?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Question about side effects and a strange tests.

File: HomeWork_5/oop_1.py (hunk 3 above)
Comment (ivanzuev78) [reply HomeWork_5/oop_1.py]: Done

File: HomeWork_5/test_oop_1.py (hunk 7 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py]: Thanks for your explanations. Done.

File: HomeWork_5/test_oop_1.py (hunk 9 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py:46]: As I understand, `student.do_homework(homework)` takes `homework` and can change this object globally. And we should check, that homework wasn't changed.

File: HomeWork_5/test_oop_1.py (hunk 8 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py]: I suppose it can modify a mutable argument passed by reference.

File: HomeWork_5/test_oop_1.py (hunk 9 above)
Comment (Cjkjvfnby) [reply HomeWork_5/test_oop_1.py:46]: > And we should check, that homework wasn't changed.

I don't think that it is required. 

You create an object with the `copy.copy`
```python
hw_copy = copy.copy(homework)
```

After that, you test that object created by `copy. copy` is a copy of the initial object.   This looks like testing `copy.copy`

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_5/test_oop_1.py (hunk 9 above)
Comment (ivanzuev78) [reply HomeWork_5/test_oop_1.py]: Yes. I understand now. And agree with you.

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**